### ISSUE TITLE
JMS Broker fails to start in LOCAL mode with JDK 11 /FISH-28

### ISSUE THREAD

**Issue by marcreichman** [payara/Payara#4551]
We typically run Glassfish with JMS in LOCAL mode for the ability to set VM parameters (e.g. flow control mode) and also to separate the memory space from the main Glassfish JVM instance, and get higher throughput. I'm porting an application from JDK 8 + Payara 5.183 to JDK 11 + Payara 5.201 and the sever will not start with the JMS service configured for LOCAL mode. I am somewhat sure it's because of the AdoptOpenJDK bundling, but I'm not sure if this would apply to other JDK 11 JDK installations as well.

I have a log attached of creating a new domain, starting it, setting the local option, and then restarting it in verbose mode to demonstrate the failure, in that the OpenMQ broker will not start because of an assumption in the JDK/JRE directory layout of the past.

Thanks for your time, and great job getting any of this going in JDK 11!

## Environment ##

- **Payara Version**: 5.201
- **Edition**: Full
- **JDK Version**: AdoptOpenJDK 11.0.6
- **Operating System**: Windows
[local-jms-fail.log](https://github.com/payara/Payara/files/4289908/local-jms-fail.log)

**Comment by rdebusscher** [comment]
Hi Marc,

There is no JRE packaging anymore in JDK11+.

But when I try the following commands, the OpenMQ broker does start.

>java -version
openjdk version "11.0.4" 2019-07-16 LTS
OpenJDK Runtime Environment Zulu11.33+15-CA (build 11.0.4+11-LTS)
OpenJDK 64-Bit Server VM Zulu11.33+15-CA (build 11.0.4+11-LTS, mixed mode)

>echo $JAVA_HOME
/Library/Java/JavaVirtualMachines/zulu-11.jdk/Contents/Home

>./asadmin start-domain
./asadmin set configs.config.server-config.jms-service.type=LOCAL
./asadmin start-domain
./asadmin stop-domain

And broker is up with JDK11

<img width="1411" alt="Screenshot 2020-03-05 at 11 22 25" src="https://user-images.githubusercontent.com/1228265/75972434-b8d07100-5ed3-11ea-9aac-eeedb4f65e81.png">

So I'm not sure what is different in your environment.

Regards
Rudy

**Comment by marcreichman** [comment]
Honestly I don’t know! I don’t have JAVA_HOME set in my environment because
I handle several different JVM installs for different projects. For this
particular setup I am using the asenv.bat change to set AS_JAVA which is
what I always do and have done. The application server itself runs just
fine with this setup, but it seems like something in the JMSRA is looking
for a “jre” directory under the main JDK folder. I did see there may be
some other environment variables I may be able to play with to influence
that which I will try today.

**Comment by marcreichman** [comment]
Can someone please try to reproduce it as I have above in Windows with AdoptOpenJDK installer? I am searching for environment variables and openmq code but it might be nice to have it confirmed if possible.

**Comment by marcreichman** [comment]
I figured out a workaround, although it's a bit ugly. The DAS by default launches the imqbrokerd with the argument `-javahome <path to java>` where the path is as determined by JAVA_HOME, PATH, or by the AS_JAVA variable like I use. Unfortunately, when the broker starts, at least on windows (I can't verify otherwise), it appends on `\jre` when using this approach, as checked by reading the openMQ source.

The OpenMQ tools also take an argument called `-jrehome` which goes right to the directory and doesn't append anything. I was able to leverage start-args in domain.xml to pass in this extra path and now my broker starts in LOCAL mode.

Here is the portion of domain.xml:

```xml
<jms-service default-jms-host="default_JMS_host" start-args="-jrehome &quot;C:\\Program Files\\AdoptOpenJDK\\jdk-11.0.6.10-hotspot&quot;" type="LOCAL">
        <jms-host port="${JMS_PROVIDER_PORT}" name="default_JMS_host" host="localhost"></jms-host>
      </jms-service>
```

If there's a better fix out there, I'd love to know!

**Comment by rdebusscher** [comment]
Hi,

I have created the following issues

Within the OpenMQ repo so that it can be fixed upstream with the JDK 11 support which is planned. https://github.com/eclipse-ee4j/openmq/issues/498

Within our repo so that we can try to fix it in the mean time, internal ticket is `CUSTCOM-222`

Regards
Rudy

**Comment by marcreichman** [comment]
Thanks very much!

**Comment by rdebusscher** [comment]
This is fixed in version 5.2021.9 (November 2021)

**Comment by marcreichman** [comment]
Thanks @rdebusscher !

**Comment by bleipold** [comment]
This fix prevents us from upgrading to Payara > 5.2021.8 on SLES 12.5. The enclsoing in quotation marks of the `jrehome` parameter in `ActiveJmsResourceAdapter.buildStartArgsForJREHome` (line 1183) causes the script `PAYARA_HOME/mq/lib/imqinit` to fail. This is because it cannot find a valid Java installation. A workaround is to remove the surrounding quotation marks in that script.

The affected part of the script:
```shell
if [ ! -x "$imq_javahome/bin/java" ]
then
   echo "Error: Invalid J2SE location: $imq_javahome"
   echo "J2SE location was specified using: $imq_javareason"
   echo "Exiting"
   exit 1
fi
```

`server.log` will contain a line like: `Error: Invalid J2SE location: "/path/to/java"` (that is just logged at INFO level btw). In fact, if really no Java installation can be found, it has to be `Error: Invalid J2SE location: /path/to/java` (without the quotation marks).

**Comment by grommitz** [comment]
I have same issue when upgrading from the 5.2021.7 to 5.2022.2 using the jdk11 docker image, which is ubuntu based. @rdebusscher do you know what has happened, you said it was fixed as of 5.2021.9?

![image](https://user-images.githubusercontent.com/2547666/181545638-5810f184-0988-4493-8b31-0e48b138cb89.png)

**Comment by JamesHillyard** [comment]
Hi @grommitz,

You are correct, this was fixed in Payara Community 5.2021.9 so if you are still experiencing this issue and believe this to be a bug, I would recommend raising a new issue with a full, easy to follow reproducer where I will be able to reinvestigate this issue and evaluate if this issue is present again in the latest version of Payara Community.

Thanks,
James

**Comment by grommitz** [comment]
Thanks James I'll try to do that at some point. I think it is probably because we do "set configs.config.server-config.jms-service.start-args="...." during the initialisation, tho we don't set -javahome in there.

I worked around this by taking a copy of the imqinit script, commenting out lines 295-311 where it tries to read -javahome and I hardcode the path instead:

imq_javahome=/usr/lib/jvm/zulu11-ca-amd64

then overwrite this file in my Dockerfile.
